infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationAckReader.java

```diff
@@ -111,43 +114,199 @@ private long getMaxReplicatedStreamsTail() {
+        }
+
+        // In Log Entry Sync
+        return calculateRemainingEntriesIncrementalUpdates(ackedTimestamp, txStreamTail, currentTxStreamProcessedTs);
+    }
+
+    private long getTxStreamTail(Map<UUID, Long> tailMap) {
+        if (tailMap.containsKey(TRANSACTION_STREAM_ID)) {
+            return tailMap.get(TRANSACTION_STREAM_ID);
+        }
+
+        log.warn("Tx Stream tail not present in sequencer, id={}", TRANSACTION_STREAM_ID);
+        return Address.NON_ADDRESS;
+    }
+
+    /**
+     * For the given replication runtime, query max stream tail for all streams to be replicated.
+     *
+     * @return max tail of all streams to be replicated for the given runtime
+     */
+    private long getMaxReplicatedStreamsTail(Map<UUID, Long> tailMap) {
+        long maxTail = Address.NON_ADDRESS;
+        for (String streamName : config.getStreamsToReplicate()) {
+            UUID streamUuid = CorfuRuntime.getStreamID(streamName);
+            if (tailMap.containsKey(streamUuid)) {
+                long streamTail = tailMap.get(streamUuid);
+                maxTail = Math.max(maxTail, streamTail);
+            }
+        }
+        return maxTail;
+    }
+
+    /**
+     * The calculation of remaining entries during log entry sync takes into account:
+     * - lastAckedTimestamp: last timestamp of entry for which we've received an ACK
+     * - txStreamTail: transaction stream tail
+     * - currentTxStreamProcessedTs: metadata of current entry being processed by the StreamsLogEntryReader,
+     *           it contains the timestamp and a boolean indicating if that entry has data to replicate or not,
+     *           this will give an indication whether if an ACK is expected or not.
+     *
+     * Consider the following representation of the data log, where entries 20, 50, 60, 70 belong to the
+     * tx stream.
+     *
+
+     +---+---+---+---+---+---+---+---+---+---+---+---+
+     |   |   |   |   |   |   |   |   |   |   |   |   |
+     +---+---+---+---+---+---+---+---+---+---+---+---+
+     20          50          60       70         100
+     tx          tx          tx       tx        (log tail / not part of tx stream)
+
+
+     * Case 1.0: Log Entry Sync lagging behind (in processing) current processing not acked with entries to replicate
+     *          - lastAckedTimestamp = 50
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 60, true (contains replicated streams)
+     *
+     *          remainingEntries = entriesBetween(50, 70] = 2
+
+     * Case 1.1: Log Entry Sync lagging behind (in processing) current processing no entries to replicate
+     *          - lastAckedTimestamp = 50
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 60, false (does not contain streams to replicate)
+     *
+     * (despite the current processed not requiring ACk, there might be entries between lastAcked and currentProcessed
```

Comment:
shouldnt this be '... entries between txStreamTail and currentTxStreamProcessed'?